File: app/src/main/java/com/dumuzeyn/mp3player/MiniPlayerRetentionPolicy.java
package com.dumuzeyn.mp3player;

/** Defines how long a stopped or paused session remains visible in the mini-player. */
public final class MiniPlayerRetentionPolicy {
    private MiniPlayerRetentionPolicy() {
    }

    public static boolean isPlaybackActive(boolean playWhenReady, PlaybackPhase phase,
            StopReason stopReason) {
        return playWhenReady
                && phase != PlaybackPhase.ENDED
                && stopReason != StopReason.QUEUE_ENDED;
    }

    static boolean isExpired(boolean playing, long inactiveSince, long legacySavedAt,
            long now, long retentionMs) {
        if (playing) {
            return false;
        }
        if (retentionMs <= 0L) {
            return true;
        }
        long startedAt = inactiveSince > 0L ? inactiveSince : legacySavedAt;
        return startedAt <= 0L || now - startedAt > retentionMs;
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/MiniPlayerRetentionPolicy.kt
package com.dumuzeyn.mp3player

/** Defines how long a stopped or paused session remains visible in the mini-player. */
object MiniPlayerRetentionPolicy {
    @JvmStatic
    fun isPlaybackActive(
        playWhenReady: Boolean,
        phase: PlaybackPhase,
        stopReason: StopReason,
    ): Boolean = playWhenReady && phase != PlaybackPhase.ENDED && stopReason != StopReason.QUEUE_ENDED

    @JvmStatic
    fun isExpired(
        playing: Boolean,
        inactiveSince: Long,
        legacySavedAt: Long,
        now: Long,
        retentionMs: Long,
    ): Boolean {
        if (playing) return false
        if (retentionMs <= 0L) return true
        val startedAt = inactiveSince.takeIf { it > 0L } ?: legacySavedAt
        return startedAt <= 0L || now - startedAt > retentionMs
    }
}
